processing/src/main/java/io/druid/segment/IndexMerger.java

```diff
@@ -1133,92 +1143,60 @@ public void createIndexDrdFile(
+    for (int i = 0; i < adapters.size(); i++) {
+      IntBuffer dimConversion = dimConversions.get(i).get(dimension);
+      if (dimConversion != null) {
+        seekers[i] = new WithConversion((IntBuffer) dimConversion.asReadOnlyBuffer().rewind());
+      } else {
+        Indexed<String> dimValueLookup = adapters.get(i).getDimValueLookup(dimension);
+        seekers[i] = new WithoutConversion(dimValueLookup == null ? 0 : dimValueLookup.size());
+      }
+    }
+    return seekers;
+  }
 
-    private int currIndex;
-    private String lastVal = null;
-    private String currValue;
+  static interface IndexSeeker
+  {
+    int NOT_EXIST = -1;
+    int NOT_INIT = -1;
 
-    DimValueConverter(
-        Indexed<String> dimSet
-    )
-    {
-      this.dimSet = dimSet;
-      conversionBuf = ByteBuffer.allocateDirect(dimSet.size() * Ints.BYTES).asIntBuffer();
+    int seek(int dictId);
+  }
 
-      currIndex = 0;
-      currValue = null;
-    }
+  static class WithoutConversion implements IndexSeeker
+  {
+    private final int limit;
 
-    public void convert(String value, int index)
+    public WithoutConversion(int limit)
     {
-      if (dimSet.size() == 0) {
-        return;
-      }
-      if (lastVal != null) {
-        if (value.compareTo(lastVal) <= 0) {
-          throw new ISE("Value[%s] is less than the last value[%s] I have, cannot be.", value, lastVal);
-        }
-        return;
-      }
-      if (currValue == null) {
-        currValue = dimSet.get(currIndex);
-      }
-
-      while (currValue == null) {
-        conversionBuf.position(conversionBuf.position() + 1);
-        ++currIndex;
-        if (currIndex == dimSet.size()) {
-          lastVal = value;
-          return;
-        }
-        currValue = dimSet.get(currIndex);
-      }
-
-      if (Objects.equal(currValue, value)) {
-        conversionBuf.put(index);
-        ++currIndex;
-        if (currIndex == dimSet.size()) {
-          lastVal = value;
-        } else {
-          currValue = dimSet.get(currIndex);
-        }
-      } else if (currValue.compareTo(value) < 0) {
-        throw new ISE(
-            "Skipped currValue[%s], currIndex[%,d]; incoming value[%s], index[%,d]", currValue, currIndex, value, index
-        );
-      }
+      this.limit = limit;
     }
 
-    public IntBuffer getConversionBuffer()
+    @Override
+    public int seek(int dictId)
     {
-      if (currIndex != conversionBuf.limit() || conversionBuf.hasRemaining()) {
-        throw new ISE(
-            "Asked for incomplete buffer.  currIndex[%,d] != buf.limit[%,d]", currIndex, conversionBuf.limit()
-        );
-      }
-      return (IntBuffer) conversionBuf.asReadOnlyBuffer().rewind();
+      return dictId < limit ? dictId : NOT_EXIST;
     }
   }
 
   /**
    * Get old dictId from new dictId, and only support access in order
    */
-  public static class DictIdSeeker
+  static class WithConversion implements IndexSeeker
```

Comment:
Can you rename "WithConversion", e.g. ConvertingIndexSeeker? Currently it sounds a bit like a boolean parameter and it's not immediately clear that it's a seeker